File: 876.Middle-of-the-Linked-List.py
# https: // leetcode.com/problems/middle-of-the-linked-list/description/
#
# algorithms
# Easy (69.1%)
# Total Accepted:    6.1k
# Total Submissions: 8.8k
# beats 9.07% of python submissions

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head.next:
          return head

        tail = head.next
        while tail:
          head = head.next
          if not tail.next or not tail.next.next:
            break
          tail = tail.next.next

        return head

File: 876.Middle-of-the-Linked-List.java
// https://leetcode.com/problems/middle-of-the-linked-list/
//
// algorithms
// Easy (66.62%)
// Total Accepted: 105,963
// Total Submissions: 160,558
// beats 100.0% of java submissions


/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode slow = new ListNode(-1);
        slow.next = head;
        ListNode fast = slow;

        while (fast != null) {
            fast = fast.next;
            if (fast != null) {
                fast = fast.next;
            }

            slow = slow.next;
        }

        return slow;
    }
}